File: solutions/232.Implement_Queue_using_Stacks/AC_front_back_stack_1.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_front_back_stack_1.cpp
*  Create Date: 2015-07-27 09:41:49
*  Descripton:   
*/


#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Queue {
private:
    stack<int> stack_front;
    stack<int> stack_back;
    void back_to_front() {
        while (!stack_back.empty()) {
            stack_front.push(stack_back.top());
            stack_back.pop();
        }
    }

public:
    // Push element x to the back of queue.
    void push(int x) {
        stack_back.push(x);
    }

    // Removes the element from in front of queue.
    void pop(void) {
        if (stack_front.empty()) {
            back_to_front();
        }
        return stack_front.pop();
    }

    // Get the front element.
    int peek(void) {
        if (stack_front.empty()) {
            back_to_front();
        }
        return stack_front.top();
    }

    // Return whether the queue is empty.
    bool empty(void) {
        return stack_front.empty() && stack_back.empty();
    }
};

int main() {

    return 0;
}

File: solutions/232.Implement_Queue_using_Stacks/AC_front_back_stack_1.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_front_back_stack_1.py
# Create Date: 2015-07-27 09:52:54
# Usage:       AC_front_back_stack_1.py 
# Descripton:  


class Queue:
    # initialize your data structure here.
    def __init__(self):
        self.front = []
        self.back = []
    
    def move_back_to_front(self):
        while self.back:
            self.front.append(self.back[-1])
            self.back.pop()

    # @param x, an integer
    # @return nothing
    def push(self, x):
        self.back.append(x)

    # @return nothing
    def pop(self):
        if not self.front:
            self.move_back_to_front()
        self.front.pop()

    # @return an integer
    def peek(self):
        if not self.front:
            self.move_back_to_front()
        return self.front[-1]

    # @return an boolean
    def empty(self):
        return not self.back and not self.front
        

# test
q = Queue()
q.push(1)
print q.empty()
print q.peek()